usine.py

```diff
@@ -1,23 +1,349 @@
+from io import StringIO
+from getpass import getuser
+from hashlib import md5
+from pathlib import Path
 
-connection = None
+from paramiko.client import SSHClient, WarningPolicy
+from paramiko.config import SSHConfig
+from progressist import ProgressBar
+import yaml
 
+client = None
 
-def init(host):
-    connection = Channel(host)
-    connection.context = {}
+
+def gray(s):
+    return f'\x1b[1;30m{s}\x1b[0m'
+
+
+def red(s):
+    return f'\x1b[1;41m{s}\x1b[0m'
+
+
+class RemoteError(Exception):
+    pass
+
+
+class Config:
+
+    def __init__(self, value=None):
+        if isinstance(value, Config):
+            value = value.value
+        super().__setattr__('value', value or {})
+
+    def get(self, key, default=None):
+        try:
+            return Config(self.value.get(key, default))
+        except AttributeError:
+            return Config(default)
+
+    def __getitem__(self, key):
+        return self.__getattr__(key)
+
+    def __getattr__(self, key):
+        try:
+            return Config(self.value.get(key))
+        except AttributeError:
+            return Config()
+
+    def __setattr__(self, key, value):
+        self.value[key] = value
+
+    def __str__(self):
+        return str(self.value)
+
+    def __eq__(self, other):
+        return other == self.value
+
+    # https://eev.ee/blog/2012/03/24/python-faq-equality/
+    # No way to override "is" behaviour, so no way to do "config.key is None"…
+    def __bool__(self):
+        return bool(self.value)
+
+    def __hash__(self):
+        return hash(self.value)
+
+    def __iter__(self):
+        return iter(self.value)
+
+    def items(self):
+        return self.value.items()
+
+    def update(self, other):
+        self.value.update(other)
+
+
+config = Config()  # singleton.
+
+
+class Template(string.Template):
+    # Default delimiter ($) clashes at least with Nginx DSL.
+    delimiter = '$$'
+
+
+def template(path, **context):
+    with Path(path).open() as f:
+        template = Template(f.read())
+        return StringIO(template.substitute(**context))
+
+
+class Formatter(string.Formatter):
+    """
+    Allow to have some custom formatting types.
+
+    B: boolean attribute
+    S: small boolean attribute
+    V: k=v like attribute
+    """
+
+    def _vformat(self, format_string, args, kwargs, used_args, recursion_depth,
+                 auto_arg_index=0):
+        result = []
+        for literal_text, name, spec, conversion in self.parse(format_string):
+
+            # output the literal text
+            if literal_text:
+                result.append(literal_text)
+
+            # if there's a field, output it
+            if name:
+                # given the name, find the object it references
+                #  and the argument it came from
+                obj, _ = self.get_field(name, args, kwargs)
+                obj = self.convert_field(obj, conversion)
+
+                value = ''
+                if spec == 'bool':
```

Comment:
I wonder if we should extract that part into a dedicated method in order to ease subclassing (additional specs).